File: solutions/cpp/1.two-sum.cpp
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> m;
        for (int i = 0; i < nums.size(); ++i) {
            int diff = target - nums[i];
            if (m.count(nums[i])) {
                return {m[nums[i]], i};
            } else {
                m[diff] = i;
            }
        }
        return {};
    }
};

void printVector(const vector<int>& v) {
   for (int i=0; i < v.size(); ++i) cout << v.at(i) << ' ';
   cout << endl;
   return;
}

int main() {
    Solution soln;
    vector<int> nums = {2, 7, 11, 15};
    int target = 9;
    vector<int> res = soln.twoSum(nums, target);
    cout << "nums: ";
    printVector(nums);
    cout << "target: " << target << endl;
    cout << "result: ";
    printVector(res);
    return 0;
}

File: solutions/python/1.two-sum.py
# 1. Two Sum

from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        diff = {}
        for i in range(len(nums)):
            if nums[i] not in diff:
                diff[target-nums[i]] = i
            else:
                return diff[nums[i]], i


def test():
    s = Solution()

    nums = [2, 7, 11, 15]
    target = 9
    result = s.twoSum(nums, target)
    print(f'{nums=}\n{target=}\n{result=}')


if __name__ == '__main__':
    test()
